### PR TITLE
Implement paddle frontend methods for partial conversion

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Implement methods for partial conversion*

### Tickets:
 - *52265*

### CODE REVIEW

File: ngraph/frontend/paddlepaddle/src/frontend.cpp [hunk 1]
```diff
@@ -35,43 +34,101 @@ namespace ngraph
...
+                        if (it != nodes.end())
+                        {
+                            inputs_vector.push_back(it->second);
+                            inputs_names.push_back(in_tensor_name);
+                        }
+                        else
+                        {
+                            // return empty map when not all inputs exist. It usually means that
+                            // these nodes are not used because model inputs were overwritten
+                            return named_outputs;
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/frontend.cpp:88]: Looks like this is not covered by tests

File: ngraph/frontend/paddlepaddle/src/frontend.cpp [hunk 2]
```diff
@@ -288,10 +354,92 @@ namespace ngraph
...
+                [&](const std::map<std::string, Output<Node>>& nodes_dict,
+                    const std::shared_ptr<OpPlacePDPD>& op_place) {
+                    pdpd::NamedOutputs named_outputs;
+                    try
+                    {
+                        named_outputs = pdpd::make_ng_node(nodes_dict, op_place, CREATORS_MAP);
+                    }
+                    catch (const OpConversionFailure&)
+                    {
+                        named_outputs = pdpd::make_framework_node(nodes_dict, op_place);
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/frontend.cpp:403]: This is also not covered by tests, and due to this `convert_partially` doesn't look fully tested

File: ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            return m_decoder.map_for_each_input([&](std::string name) {
+                auto it = std::find(m_inputs_names.begin(), m_inputs_names.end(), name);
+                if (it != m_inputs_names.end())
+                {
+                    return input(it - m_inputs_names.begin()).get_source_output();
+                }
+                else
+                {
+                    return Output<Node>();
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp:23]: minor: This line not covered by unit tests

File: ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                }
+            });
+        }
+
+        std::map<std::string, OutputVector> PDPDFrameworkNode::get_named_outputs()
+        {
+            size_t idx = 0;
+            return m_decoder.map_for_each_output([&](std::string name) { return output(idx++); });
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp:31]: should be 'const std::string &'

File: ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                    return input(it - m_inputs_names.begin()).get_source_output();
+                }
+                else
+                {
+                    return Output<Node>();
+                }
+            });
+        }
+
+        std::map<std::string, OutputVector> PDPDFrameworkNode::get_named_outputs()
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp:28]: Is there really no way to make it 'const'?

File: ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp (hunk 4 above)
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp:31]: Is `name` really not needed here?

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp [hunk 6]
```diff
@@ -40,7 +67,10 @@ namespace ngraph
 
         private:
             static std::shared_ptr<Function>
-                convert_model(const std::shared_ptr<InputModelPDPD>& model);
+                convert_each_node(const std::shared_ptr<InputModelPDPD>& model,
```
Comment (itikhono) [inline ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp:70]: can we completely move this function to .cpp file?

File: ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp (hunk 4 above)
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp:31]: `idx `should be always 0?

File: ngraph/frontend/paddlepaddle/src/decoder.cpp [hunk 7]
```diff
@@ -135,5 +145,40 @@ namespace ngraph
                                     " Expected number: 0 or 1");
             return attrs;
         }
+
+        std::map<std::string, OutputVector> DecoderPDPDProto::map_for_each_input(
+            std::function<Output<Node>(const std::string&)> func) const
```
Comment (ilyachur) [inline ngraph/frontend/paddlepaddle/src/decoder.cpp:150]:
```suggestion
            const std::function<Output<Node>(const std::string&)>& func) const
```

File: ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp (hunk 4 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp:31]: > idx should be always 0?
Yes, it starts from 0.

Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp:31]: > Is name really not needed here?

Yes it is not needed here, but I decided to make it in a similar way as inputs.

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp (hunk 6 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp:70]: Unfortunately, this function needs access to protected members of InputModelPDPD, so it has to be part of frontend.

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp [hunk 8]
```diff
@@ -22,6 +23,32 @@ namespace ngraph
...
+                convert(std::shared_ptr<ngraph::Function> partiallyConverted) const override;
+
+            /// \brief Convert only those parts of the model that can be converted leaving others
+            /// as-is. Converted parts are not normalized by additional transformations; normalize
+            /// function or another form of convert function should be called to finalize the
+            /// conversion process.
+            /// \param model Input model
+            /// \return partially converted nGraph function
+            std::shared_ptr<ngraph::Function>
+                convert_partially(InputModel::Ptr model) const override;
```
[the remark this answers is not in the document]
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp:39]: those are pointers and all API is written this way, should we change it now?

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp [hunk 9]
```diff
@@ -22,6 +23,32 @@ namespace ngraph
             /// \return fully converted nGraph function
             std::shared_ptr<Function> convert(InputModel::Ptr model) const override;
 
+            /// \brief Completely convert the remaining, not converted part of a function.
+            /// \param partiallyConverted partially converted nGraph function
+            /// \return fully converted nGraph function
+            std::shared_ptr<ngraph::Function>
+                convert(std::shared_ptr<ngraph::Function> partiallyConverted) const override;
```
[the remark this answers is not in the document]
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp:30]: Thats how API was written, should we change it?

[the remark this answers is not in the document]
Comment (ilyachur) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp:30]: I propose yes, we should change it.

File: ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace ngraph
+{
+    namespace frontend
+    {
+        NGRAPH_RTTI_DEFINITION(PDPDFrameworkNode, "PDPDFrameworkNode", 1);
+
+        std::map<std::string, OutputVector> PDPDFrameworkNode::get_named_inputs() const
+        {
+            return m_decoder.map_for_each_input([&](std::string name) {
```
Comment (mvafin) [inline ngraph/frontend/paddlepaddle/src/pdpd_fw_node.cpp:15]: Pass const ref

File: ngraph/frontend/paddlepaddle/src/pdpd_fw_node.hpp [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+                return std::make_shared<PDPDFrameworkNode>(m_decoder, inputs, m_inputs_names);
+            }
+
+            std::string get_op_type() const { return m_decoder.get_op_type(); }
+
+            const DecoderPDPDProto& get_decoder() const { return m_decoder; }
+
+            std::map<std::string, OutputVector> get_named_inputs() const;
+
+            std::map<std::string, OutputVector> get_named_outputs();
```
Comment (mvafin) [inline ngraph/frontend/paddlepaddle/src/pdpd_fw_node.hpp:42]: rename to return_named_outputs

**Review by nosovmik — APPROVED** [review]
Looks good to me. Hopefully nobody will create 'rxyz' operation :-)

**Review by itikhono — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Pull request merged**